app/services/articles/feeds/weighted_query_strategy.rb

```diff
@@ -441,6 +442,28 @@ def featured_story_and_default_home_feed(**)
 
       private
 
+      def final_order_logic(articles)
+        case @strategy
+        when "final_order_by_score"
+          articles.order("score DESC")
+        when "final_order_by_comment_score"
+          articles.order("comment_score DESC")
+        when "final_order_by_last_comment_at"
+          articles.order("articles.last_comment_at DESC")
+        when "final_order_by_random"
+          articles.order("RANDOM()")
+        when "final_order_by_random_weighted_to_score"
+          articles.order(Arel.sql("RANDOM() ^ (1.0 / (articles.score + 1)) DESC"))
+        when "final_order_by_random_weighted_to_comment_score"
+          articles.order(Arel.sql("RANDOM() ^ (1.0 / (articles.comment_score + 1)) DESC"))
```

Comment:
What is the purpose of the +1 here and above? Is it because the value can be too low and yield a different result?

Reply:
`articles.score` can be `0`, so it's a pretty basic "don't divide by zero" hack — and I think it's fine given that this is not a very _precise_ calculation. We're just looking to take something that is _somewhat_ ordered but also somewhat randomized. 🙂

Reply:
Whoops, I actually realized this will break with any articles where the score is `-1` — need to account for this.

Reply:
_Technically_ scores of `-1` shouldn't really make it to this list in the first place, but let me rethink this just to be a little more clever.

Reply:
I don't understand this to great length but with the explanation you've given, I'm getting a feeling that it might not make sense to add `articles.score` or `articles.comment_score` into this ab testing. If the score is really varied, then it'll just look like you ordered it by said score. If you throw in a dynamic offset you may have a heavily skewed or inconsistent data.

Reply:
@maestromac The way this works is that _the initial algorithm_ takes a bunch of things into consideration: Tags followed, etc.

The _final ordering_ that we are now testing is "given the 25 posts that we came up with in the initial query, how might we best choose the display order". So in order to get into the final ordering game, posts must first win on a variety of metrics which are tested independently.

That is to say that no single post is able to overwhelm the algorithm on popularity alone.... But once it comes down to the final 25, then it _may_ be able to stick around on popularity (or randomness, or etc. based on these variable options)... So I think everything you are bringing up is accounted for, and outcomes of user preferences will ultimately be what we look at this time around.

Anyway... To account for aforementioned edge cases, I just changed the query to be `greatest(articles.score, 0.1)` which should provide the same results, for all intents and purposes, but do away with that problematic edge case. 🙂

Reply:
sounds good!